### PR TITLE
add a "test" mode

### PR DESCRIPTION
**Pull request by nvdk**
this mode works differently from dev in that it creates a merged package.json file, so we have both the dependencies of the template and of the app available. In case of overlapping keys the values from the app are preferred over those from the template. This is typically a requirement to successfully run tests, and may also be considered as a sane approach for the dev script.

Merging of the package.json files is done using jq's multiply function and probably has some limitations I'm currently not aware of. So far it seems to be working.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -2,7 +2,7 @@ FROM node:14.16.1-alpine
 
 LABEL maintainer="madnificent@gmail.com"
 
-RUN apk update && apk upgrade && apk add --no-cache bash git openssh
+RUN apk update && apk upgrade && apk add --no-cache bash git jq
```
Comment (nvdk) [inline Dockerfile:5]: removed ssh since I don't really see a usecase for it, but can add again if necessary?

File: boot.sh [hunk 2]
```diff
@@ -3,8 +3,11 @@ then
     # Run live-reload development
     exec /usr/src/app/node_modules/.bin/nodemon \
          --watch /app \
-         --ext js,mjs,cjs,json \
+         --ext js,mjs,cjs,json,hbs \
```
Comment (nvdk) [inline boot.sh:6]: not really related to this PR, but reload on changed hbs files is appreciated

File: run-test.sh [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
+
+# create a copy of app, so we can modify as necessary
+cp -a /app /usr/src/out
+
+# add some required files from the template
+cp -a /usr/src/app/.babelrc.js /usr/src/app/helpers /usr/src/out/
+# merge package.json files, prefering whatever is specified in the app
+jq -s '.[0] * .[1]' /usr/src/app/package.json  /app/package.json > /usr/src/out/package.json
```
Comment (nvdk) [inline run-test.sh:9]: `-s ` is for slurp, which means read the entire thing before applying the operation
`*`  is the multiplier operation, documented as:
> Multiplying two objects will merge them recursively: this works like addition but if both objects contain a value for the same key, and the values are objects, the two are merged with the same strategy.

**Comment by nvdk** [comment]
this still has an issue where the shell doesn't quit after running the tests, haven't figured out why yet.

**Comment by madnificent** [comment]
Hi @nvdk.  I would like to let this issue blow up :-)

This seems to be strongly related to the dev setup issue you have with the 'mu' package in this package as well as in the mu-semtech/mu-ruby-template package.  Your aim was to have a dev setup that would also be installable by regular scripts.

My understanding is that a published package would help solve some issues.  With the slowly evolving branding to semantic.works, we could make a shift to published packages perhaps.  A first ideation leads to publishing a 'semantic-works' npm package from which the current imports can happen (`query`, `app`, `sparqlEscapeString`, ...).  The template may depend on more than only that package, hence we could publish a package that contains all dependencies in lockstep with the template's version.

Such approach would mean we can execute this merge on all environments.  For backwards compatibility, we could keep supporting the current mu imports as well, though we might require an upgrade if you want to use this tests feature.

Does this approach sound reasonable?  Is ticket more urgent right now?  If you expect further discussion on this topic specifically, I suggest we move the discussion into a separate issue on mu-semtech/mu-project.

**Comment by nvdk** [comment]
We're currently running on a fork of the javascript template that has this feature, so there is no urgency anymore. I do think you could still see things separately (that is if the template should offer testing capabilities). I agree the discussion on the future templates would fit better as an issue mu-project.

**Comment by erikap** [comment]
Replaced by #29, #43 and #44 .

**Pull request closed without merging**